frame/support/src/storage/types/counted_map.rs

```diff
@@ -0,0 +1,969 @@
+	pub fn try_mutate_exists<KeyArg, R, E, F>(key: KeyArg, f: F) -> Result<R, E>
+	where
+		KeyArg: EncodeLike<Key> + Clone,
+		F: FnOnce(&mut Option<Value>) -> Result<R, E>,
+	{
+		<Self as Helper>::Map::try_mutate_exists(key, |option_value| {
+			let existed = option_value.is_some();
+			let res = f(option_value);
+			let exist = option_value.is_some();
+
+			if res.is_ok() {
+				if existed && !exist {
+					// Value was deleted
+					CounterFor::<Prefix>::mutate(|value| value.saturating_dec());
+				} else if !existed && exist {
+					// Value was added
+					CounterFor::<Prefix>::mutate(|value| value.saturating_inc());
+				}
+			}
+			res
+		})
+	}
+
+	/// Take the value under a key.
+	pub fn take<KeyArg: EncodeLike<Key> + Clone>(key: KeyArg) -> QueryKind::Query {
+		let removed_value = <Self as Helper>::Map::mutate_exists(
+			key,
+			|value| core::mem::replace(value, None)
+		);
+		if removed_value.is_some() {
+			CounterFor::<Prefix>::mutate(|value| value.saturating_dec());
+		}
+		<Self as Helper>::Map::from_optional_value_to_query(removed_value)
+	}
+
+	/// Append the given items to the value in the storage.
+	///
+	/// `Value` is required to implement `codec::EncodeAppend`.
+	///
+	/// # Warning
+	///
+	/// If the storage item is not encoded properly, the storage will be overwritten and set to
+	/// `[item]`. Any default value set for the storage item will be ignored on overwrite.
+	pub fn append<Item, EncodeLikeItem, EncodeLikeKey>(key: EncodeLikeKey, item: EncodeLikeItem)
+	where
+		EncodeLikeKey: EncodeLike<Key> + Clone,
+		Item: Encode,
+		EncodeLikeItem: EncodeLike<Item>,
+		Value: StorageAppend<Item>
+	{
+		if !<Self as Helper>::Map::contains_key(Ref::from(&key)) {
+			CounterFor::<Prefix>::mutate(|value| value.saturating_inc());
+		}
+		<Self as Helper>::Map::append(key, item)
+	}
+
+	/// Read the length of the storage value without decoding the entire value under the given
+	/// `key`.
+	///
+	/// `Value` is required to implement [`StorageDecodeLength`].
+	///
+	/// If the value does not exists or it fails to decode the length, `None` is returned. Otherwise
+	/// `Some(len)` is returned.
+	///
+	/// # Warning
+	///
+	/// `None` does not mean that `get()` does not return a value. The default value is completly
+	/// ignored by this function.
+	pub fn decode_len<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Option<usize>
+		where Value: StorageDecodeLength,
+	{
+		<Self as Helper>::Map::decode_len(key)
+	}
+
+	/// Migrate an item with the given `key` from a defunct `OldHasher` to the current hasher.
+	///
+	/// If the key doesn't exist, then it's a no-op. If it does, then it returns its value.
+	pub fn migrate_key<OldHasher: crate::hash::StorageHasher, KeyArg: EncodeLike<Key>>(
+		key: KeyArg
+	) -> Option<Value> {
+		<Self as Helper>::Map::migrate_key::<OldHasher, _>(key)
+	}
+
+	/// Remove all value of the storage.
+	pub fn remove_all() {
```

Comment:
if we want to support limit, then we need sp-io to return the number of deleted key in the overlay, once merged I will open an issue, and we can work on it when we need it

Reply:
What is limit?

Reply:
for other storage the remove_all method allow to remove only up to a certain number of key in the backend: https://substrate.dev/rustdocs/latest/frame_support/storage/types/struct.StorageMap.html#method.remove_all

Here it is not possible without changing sp-io. Because we need to count the value removed in the overlay. And https://substrate.dev/rustdocs/latest/sp_io/storage/fn.clear_prefix.html doesn't give us this information